cpu/stm32_common/periph/i2c.c

```diff
@@ -0,0 +1,453 @@
+{
+    /* 10 bit address not supported for now */
+    (void)flags;
+
+    assert(i2c != NULL);
+
+    uint16_t tick = TICK_TIMEOUT;
+
+    i2c->CR2 = 0;
+
+    DEBUG("[i2c] start: set address mode\n");
+    /* set address mode to 7-bit */
+    i2c->CR2 &= ~(I2C_CR2_ADD10);
+
+    DEBUG("[i2c] start: set slave address\n");
+    /* set slave address */
+    i2c->CR2 &= ~(I2C_CR2_SADD);
+    i2c->CR2 |= (address << 1);
+
+    DEBUG("[i2c] start: set transfert direction\n");
+    /* set transfer direction */
+    i2c->CR2 &= ~(I2C_CR2_RD_WRN);
+    i2c->CR2 |= (rw_flag << I2C_CR2_RD_WRN_Pos);
+
+    DEBUG("[i2c] start: set number of bytes\n");
+    /* set number of bytes */
+    i2c->CR2 &= ~(I2C_CR2_NBYTES);
+    i2c->CR2 |= (length << I2C_CR2_NBYTES_Pos);
+
+    /* configure autoend configuration */
+    i2c->CR2 &= ~(I2C_CR2_AUTOEND);
+
+    /* Clear interrupt */
+    i2c->ICR |= CLEAR_FLAG;
+
+    /* generate start condition */
+    DEBUG("[i2c] start: generate start condition\n");
+    i2c->CR2 |= I2C_CR2_START;
+
+    /* Wait for the start followed by the address to be sent */
+    while (!(i2c->CR2 & I2C_CR2_START) && tick--) {}
+}
+
+static inline int _read(I2C_TypeDef *i2c, uint8_t *data, int length)
+{
+    assert(i2c != NULL);
+
+    uint16_t tick = TICK_TIMEOUT;
+
+    for (int i = 0; i < length; i++) {
+        /* wait for transfer to finish */
+        DEBUG("[i2c] read: Waiting for DR to be full\n");
+        while (!(i2c->ISR & I2C_ISR_RXNE) && tick--) {
+            if (i2c->ISR & ERROR_FLAG || !tick) {
+                return -1;
+            }
+        }
+
+        DEBUG("[i2c] read: DR is now full\n");
+
+        /* read data from data register */
+        data[i] = i2c->RXDR;
+        DEBUG("[i2c] read: Read byte %i from DR\n", i);
+    }
+
+    return 0;
+}
+
+static inline int _write(I2C_TypeDef *i2c, const uint8_t *data, int length)
+{
+    assert(i2c != NULL);
+
+    uint16_t tick = TICK_TIMEOUT;
+
+    for (int i = 0; i < length; i++) {
+        /* wait for ack */
+        DEBUG("[i2c] write: Waiting for ACK\n");
+        while (!(i2c->ISR & I2C_ISR_TXIS) && tick--)  {
+            if (i2c->ISR & ERROR_FLAG || !tick) {
+                return -1;
+            }
+        }
+
+        /* write data to data register */
+        DEBUG("[i2c] write: Write byte %i to DR\n", i);
+        i2c->TXDR = data[i];
+        DEBUG("[i2c] write: Sending data\n");
+    }
+
+    return 0;
+}
+
+static inline void _stop(I2C_TypeDef *i2c)
+{
+    assert(i2c != NULL);
+
+    uint16_t tick = TICK_TIMEOUT;
+
+    /* make sure transfer is complete */
+    DEBUG("[i2c] stop: Wait for transfer to be complete\n");
+    while (!(i2c->ISR & I2C_ISR_TC) && tick--) {
+        if (i2c->ISR & ERROR_FLAG || !tick) {
+            break;
+        }
+    }
+
+    /* send STOP condition */
+    DEBUG("[i2c] stop: Generate stop condition\n");
+    i2c->CR2 |= I2C_CR2_STOP;
+}
+
+static inline void irq_handler(i2c_t dev)
```

Comment:
It seems no interrupt is enable, is this function really needed or is this more for a future use?

Reply:
One interrupt is enabled here: https://github.com/RIOT-OS/RIOT/pull/9202/files#diff-8703624eb1944c61de7e7320bf02f6e3R80

On certain families (F3, L4, F7) there are 2 per I2C (ev/er), I only took the `er` for the moment.